src/plugin/session-tracker.ts

```diff
@@ -0,0 +1,182 @@
+ * Copyright (C) 2024 EclipseSource.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { DebugProtocol } from '@vscode/debugprotocol';
+import * as vscode from 'vscode';
+import { isDebugEvent, isDebugRequest, isDebugResponse } from '../common/debug-requests';
+import { WrittenMemory } from '../common/memory-range';
+
+export interface SessionInfo {
+    raw: vscode.DebugSession;
+    debugCapabilities?: DebugProtocol.Capabilities;
+    clientCapabilities?: DebugProtocol.InitializeRequestArguments;
+    active?: boolean;
+    stopped?: boolean;
+}
+
+export interface SessionEvent {
+    event: string;
+    session?: SessionInfo;
+    data?: unknown;
+}
+
+export interface ActiveSessionChangedEvent extends SessionEvent {
+    event: 'active';
+}
+
+export interface SessionMemoryWrittenEvent extends SessionEvent {
+    event: 'memory-written';
+    session: SessionInfo;
+    data: WrittenMemory;
+}
+
+export interface SessionStoppedEvent extends SessionEvent {
+    event: 'stopped';
+    session: SessionInfo;
+}
+
+export interface SessionContinuedEvent extends SessionEvent {
+    event: 'continued';
+    session: SessionInfo;
+}
+
+export interface SessionEvents {
+    'active': ActiveSessionChangedEvent,
+    'memory-written': SessionMemoryWrittenEvent,
+    'continued': SessionContinuedEvent,
+    'stopped': SessionStoppedEvent
+}
+
+export type DebugCapability = keyof DebugProtocol.Capabilities;
+export type ClientCapability = keyof DebugProtocol.InitializeRequestArguments;
+
+export function isSessionEvent<K extends keyof SessionEvents>(event: K, message: unknown): message is SessionEvents[K] {
+    const assumed = message ? message as SessionEvent : undefined;
+    return !!assumed && assumed.event === event;
+}
+
+export class SessionTracker implements vscode.DebugAdapterTrackerFactory {
+    protected toDispose: vscode.Disposable[] = [];
+
+    protected readonly _sessionInfo = new Map<string, SessionInfo>();
+
+    private _onSessionEvent = new vscode.EventEmitter<SessionEvent>();
+    public readonly onSessionEvent = this._onSessionEvent.event;
+
+    activate(context: vscode.ExtensionContext): void {
+        context.subscriptions.push(
+            vscode.debug.registerDebugAdapterTrackerFactory('*', this),
```

Comment:
I understand the conceptual separation of jobs between this class and the memory provider, and I like the idea of separating the tracking part from the memory provider part, but I'm not thrilled about both of them registering tracker factories on `*`. It's a tricky problem: I was looking at [this plugin](https://github.com/mcu-debug/debug-tracker-vscode) earlier today, and on the one hand, I understood the interest in centralization, but on the other hand, you're likely to still to have the same number of handlers doing the same amount of work, with the only difference being which ones VSCode calls directly and which ones it calls indirectly.

I think for now, I'm basically happy with the division of labor you've implemented, but if you see a way to avoid doubling the session trackers, I might like it even better.

Reply:
Hmm... that is a very good point and I too was a bit frustrated with the lack of data that VS Code gives about the session, so I definitely see how that plugin came to be. While it centralizes the code nicely (which is a big plus imho), I don't think it will really improve performance since the amount of work will be mostly the same.

I'm not sure how we can reduce the number of session trackers without simply introducing a level of indirection which will lead to a very similar result than the plugin you mentioned. However, I'm also not sure how much impact that will have since currently we have two "global" trackers, one for session information and one for memory. If that number grows and the connections between the trackers become somehow difficult to manage, we should definitely re-think this. But if you agree, I'd postpone that for now and maybe track that in a separate issue if we encounter any problems.